src/persist-client/src/write.rs

```diff
@@ -403,57 +679,85 @@ where
         );
 
         let batch = BlobTraceBatchPart {
-            desc,
-            updates: builder.finish(),
+            desc: encoded_desc.clone(),
+            updates,
             index: 0,
         };
-        if batch.updates.len() == 0 {
-            return Ok(None);
-        }
 
         let mut buf = Vec::new();
         batch.encode(&mut buf);
-        Ok(Some(buf))
+
+        let key = Uuid::new_v4().to_string();
+        let () = retry_external("compare_and_append::set", || async {
+            // If MultiBlob::set took value as a ref, then we wouldn't have
+            // to clone here.
+            self.blob
+                .set(
+                    Instant::now() + FOREVER,
+                    &key,
+                    buf.clone(),
+                    Atomicity::RequireAtomic,
+                )
+                .await
+        })
+        .await;
+
+        let batch = Batch::new(desc, vec![key]);
+
+        Ok(batch)
     }
 
-    /// Test helper for an [Self::append] call that is expected to succeed.
-    #[cfg(test)]
-    #[track_caller]
-    pub async fn expect_append<L, U>(&mut self, updates: &[((K, V), T, D)], lower: L, new_upper: U)
+    /// Adds the given update to the batch.
+    ///
+    /// The update timestamp must be greater or equal to `lower` that was given
+    /// when creating this [BatchBuilder].
+    pub async fn add<SB, KB, VB, TB, DB>(&mut self, update: SB) -> Result<(), InvalidUsage<T>>
     where
-        L: Into<Antichain<T>>,
-        U: Into<Antichain<T>>,
+        SB: Borrow<((KB, VB), TB, DB)>,
```

Comment:
> unless I'm missing something, this shouldn't need all the Borrow params. just take a k: &K, v: &V, t: &T, d: &D and make the caller deal with it

changed!

Reply:
I think `AsRef<>` or `Borrow<>` would allow more flexibility for the caller, but I don't have a strong opinion so will go with simple references for now.

I still have this rust beginner -> rust pro thing in my head where you go:
 - `String`
 - `&str`
 - `AsRef<str>`
 - maybe then `Cow<'a, str>`

😅